### ISSUE TITLE
[hold for payment 2023-08-30] [$2000] Security - Copy button is displayed without any secret key when the 2FA step 2 page is accessed directly through URL

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#19496]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Got to https://staging.new.expensify.com/settings/security/two-factor-auth/verify on an account where you have not previously copied or downloaded 2FA secret keys. 
2. Sign in if not already.
3. Notice you are not shown the 2FA secret key when the page loads. Nor if you go back, can you see the recovery codes on the previous page. 

## Expected Result:
Here are how we want to handle all of the various cases related to this issue: 

**2FA is not enabled, and the user loads:**

- [new.expensify.com/settings/security/two-factor-auth/codes](http://new.expensify.com/settings/security/two-factor-auth/codes) > we show codes.
- [new.expensify.com/settings/security/two-factor-auth/verify](http://new.expensify.com/settings/security/two-factor-auth/verify) > we redirect to step 1 (codes).
- [new.expensify.com/settings/security/two-factor-auth/success](http://new.expensify.com/settings/security/two-factor-auth/success) > we redirect to step 1 (codes).
- [new.expensify.com/settings/security/two-factor-auth/enabled](http://new.expensify.com/settings/security/two-factor-auth/enabled) > we redirect to step 1 (codes).
- http://new.expensify.com/settings/security/two-factor-auth/disable > we redirect to step 1 (codes). 

**2FA is enabled, and the user loads:**

- [new.expensify.com/settings/security/two-factor-auth/codes](http://new.expensify.com/settings/security/two-factor-auth/codes) > we redirect to [new.expensify.com/settings/security/two-factor-auth/enabled](http://new.expensify.com/settings/security/two-factor-auth/enabled)
- [new.expensify.com/settings/security/two-factor-auth/verify](http://new.expensify.com/settings/security/two-factor-auth/verify) > we redirect to [new.expensify.com/settings/security/two-factor-auth/enabled (http://new.expensify.com/settings/security/two-factor-auth/enabled)
- [new.expensify.com/settings/security/two-factor-auth/success](http://new.expensify.com/settings/security/two-factor-auth/success) > we redirect to [new.expensify.com/settings/security/two-factor-auth/enabled](http://new.expensify.com/settings/security/two-factor-auth/enabled).
- [new.expensify.com/settings/security/two-factor-auth/enabled](http://new.expensify.com/settings/security/two-factor-auth/enabled) > they land on enabled. 
- http://new.expensify.com/settings/security/two-factor-auth/disable > they land on http://new.expensify.com/settings/security/two-factor-auth/enabled . 

## Actual Result:
Copy button is there without any secret key shown. 

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.17.0

**Reproducible in staging?:** Yes

**Reproducible in production?:** yes

**If this was caught during regression testing, add the test name, ID and link from TestRail:**

**Email or phone of affected tester (no customers):**

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/93399543/312c2fb3-972f-47ab-8fe7-009318c4c7dd

**Expensify/Expensify Issue URL:**

**Issue reported by:** @adeel0202 

**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1684420754597459

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~012f90e1d072b364da</li>
        <li>Upwork Job ID: 1661463527223799808</li>
        <li>Last Price Increase: 2023-06-14</li>
    </ul>
</details>

**Comment by joekaufmanexpensify** [comment]
I can reproduce this on web. If I had to suspect why this happening, it's because typically when enabling 2FA, we require you to download or copy the recovery codes, before proceeding the step that has the secret key. Whereas by loading https://staging.new.expensify.com/settings/security/two-factor-auth/verify , the user is bypassing this required first step. 

IMO, the best solution would be to redirect the user to the first page of the 2FA flow (to copy/download their recovery codes), if they try and load https://staging.new.expensify.com/settings/security/two-factor-auth/verify directly, and have niot previously copied or downloaded their recovery codes.

**Comment by kushu7** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
 Copy button is displayed without any secret key when the 2FA step 2 page is accessed directly through URL
### What is the root cause of that problem?
Its happening because user directly accessing `settings/security/two-factor-auth/verify` page. we are missing recoveryCodes and twoFactorAuthSecretKey.
### What changes do you think we should make in order to solve the problem?
We should redirect user to 1st step and toggle2FA.
https://github.com/Expensify/App/blob/6c23b160e2ef8d0184e8a3c574b5e117270a5085/src/pages/settings/Security/TwoFactorAuth/VerifyPage.js#L54-L59
Can add below this function.
```js
    useEffect(() => {
        if (props.account.recoveryCodes) {
            return;
        }
        Session.toggleTwoFactorAuth(true);
        Navigation.navigate(ROUTES.SETTINGS_2FA_CODES)
    }, [props.account.recoveryCodes])
```


### What alternative solutions did you explore? (Optional)
None

**Comment by joekaufmanexpensify** [comment]
@aimane-chnaif thoughts on these proposals?

**Comment by joekaufmanexpensify** [comment]
Adding an additional assignee as I'm OOO next week (until June 5th). Next steps here is to make sure we get these proposals reviewed, and hopefully select one!

**Comment by Christinadobrzyn** [comment]
@deetergp can you double check on this - https://github.com/Expensify/App/issues/19496#issuecomment-1567239748

Also, @aimane-chnaif let us know how these proposals are looking or if you'd like to see more. TY!

**Comment by deetergp** [comment]
> @deetergp should we fix this as well here? I believe yes but just to confirm.

We should definitely fix it there too @aimane-chnaif.

**Comment by Christinadobrzyn** [comment]
still waiting on proposals - 

@aimane-chnaif or @deetergp do you think we should reach out to Callstack or Maison to see if they can help with this?

**Comment by joekaufmanexpensify** [comment]
@Christinadobrzyn I'm back from OOO and can take this one back. Thanks for covering!

**Comment by arthurgubaidullin** [comment]
Contributor details
Your Expensify account email: lapteuh@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~0174ed83dbf8ccc5c3

**Comment by joekaufmanexpensify** [comment]
@aimane-chnaif Before we go the route of an agency here, could you please review these two new proposals to see if either work?

**Comment by joekaufmanexpensify** [comment]
Hm, yeah that's interesting. I agree with 1. 

For 2 and 3, I kinda like always redirecting to step 1 (if the user has not set up 2fa). The reality is it's going to be very unlikely that many/any users are ever going to be loading directly into one of the intermediate steps of setting up 2FA. 

IMO the goal here should be to ensure as simply as possible that the user always correctly sets up 2FA, as otherwise their account could become unrecoverable (if they set up 2FA and never downloaded recovery codes). And redirecting to step 1 in every situation feels simplest to me.

**Comment by joekaufmanexpensify** [comment]
Separately, a few other cases we should fix here IMO: 

- If user has not enabled 2FA and tries to load `https://new.expensify.com/settings/security/two-factor-auth/enabled`, they should be redirected to step 1 of the process to set up 2FA. 
- If user who has enabled 2FA tries to load any of the three URLs in the 2FA set up process, they should prob be redirected to https://new.expensify.com/settings/security/two-factor-auth/enabled. Right now if you load `new.expensify.com/settings/security/two-factor-auth/codes` you get the blank codes screen. And if you load `new.expensify.com/settings/security/two-factor-auth/verify` or `new.expensify.com/settings/security/two-factor-auth/success`, you land on `new.expensify.com/settings/security/two-factor-auth/success`. Which is odd as the user cannot otherwise access this screen in the product once 2FA is enabled.

**Comment by joekaufmanexpensify** [comment]
@aimane-chnaif @deetergp curious if you agree with what I outlined above?

**Comment by arthurgubaidullin** [comment]
## Proposal
[Updated](https://github.com/Expensify/App/issues/19496#issuecomment-1579162504)

**Comment by joekaufmanexpensify** [comment]
>if the user has already successfully gone through the process of enabling 2FA, we should show the codes.

@deetergp should we show the codes once 2FA is already enabled? I thought we only show the codes until the user has enabled 2fa

**Comment by deetergp** [comment]
@joekaufmanexpensify I would think we'd want them to be able to see their codes after they've enabled 2FA. I can go to my Github settings and see my codes and I've had 2FA enabled there for years.

**Comment by joekaufmanexpensify** [comment]
Started discussion [here. ](https://expensify.slack.com/archives/C01GTK53T8Q/p1686692724943929)

**Comment by bondydaa** [comment]
> I would think we'd want them to be able to see their codes after they've enabled 2FA

We purposefully decided not to allow a user to re-see these and that you can only get them during set up.

This was to make it so that someone would not be able to get access to your backup codes if you leave your computer unlocked or signed in to a public computer.

**Comment by deetergp** [comment]
> We purposefully decided not to allow a user to re-see these and that you can only get them during set up.

I must have missed out on that initial discussion, but TIHI 😏

**Comment by joekaufmanexpensify** [comment]
We are close to aligning on how we want to handle all of the various cases in [the slack thread](https://expensify.slack.com/archives/C01GTK53T8Q/p1686771041292679?thread_ts=1686692724.943929&cid=C01GTK53T8Q). As soon as we do, I will update the issue, so we can quickly look for proposals, and get a PR up!

**Comment by joekaufmanexpensify** [comment]
We aligned in the slack thread, and here's how we want to handle all of the various cases in this issue: 

**2FA is not enabled, and the user loads:**

- [new.expensify.com/settings/security/two-factor-auth/codes](http://new.expensify.com/settings/security/two-factor-auth/codes) > we show codes.
- [new.expensify.com/settings/security/two-factor-auth/verify](http://new.expensify.com/settings/security/two-factor-auth/verify) > we redirect to step 1 (codes).
- [new.expensify.com/settings/security/two-factor-auth/success](http://new.expensify.com/settings/security/two-factor-auth/success) > we redirect to step 1 (codes).
- [new.expensify.com/settings/security/two-factor-auth/enabled](http://new.expensify.com/settings/security/two-factor-auth/enabled) > we redirect to step 1 (codes).
- http://new.expensify.com/settings/security/two-factor-auth/disable > we redirect to step 1 (codes). 

**2FA is enabled, and the user loads:**

- [new.expensify.com/settings/security/two-factor-auth/codes](http://new.expensify.com/settings/security/two-factor-auth/codes) > we redirect to [new.expensify.com/settings/security/two-factor-auth/enabled](http://new.expensify.com/settings/security/two-factor-auth/enabled)
- [new.expensify.com/settings/security/two-factor-auth/verify](http://new.expensify.com/settings/security/two-factor-auth/verify) > we redirect to [new.expensify.com/settings/security/two-factor-auth/enabled (http://new.expensify.com/settings/security/two-factor-auth/enabled)
- [new.expensify.com/settings/security/two-factor-auth/success](http://new.expensify.com/settings/security/two-factor-auth/success) > we redirect to [new.expensify.com/settings/security/two-factor-auth/enabled](http://new.expensify.com/settings/security/two-factor-auth/enabled).
- [new.expensify.com/settings/security/two-factor-auth/enabled](http://new.expensify.com/settings/security/two-factor-auth/enabled) > they land on enabled. 
- http://new.expensify.com/settings/security/two-factor-auth/disable > they land on http://new.expensify.com/settings/security/two-factor-auth/enabled . 

I updated the `Expected results` section of OP to reflect this as well.

**Comment by joekaufmanexpensify** [comment]
@kushu7 @hoangzinh @WikusKriek @arthurgubaidullin tagging all of you as you posted proposals on this issue. We updated the [expected behavior](https://github.com/Expensify/App/issues/19496#issuecomment-1593600230) for how we want this to work. 

Please feel free to post new proposals with this in mind, and we'll review them quickly. We'd like to make progress on this one quickly!

**Comment by joekaufmanexpensify** [comment]
Next step here is for @aimane-chnaif to review these updated proposals!

**Comment by arthurgubaidullin** [comment]
## Proposal
[Updated](https://github.com/Expensify/App/issues/19496#issuecomment-1579162504)

**Comment by joekaufmanexpensify** [comment]
Proposals awaiting review, and @aimane-chnaif confirmed they'll be able to do this today.

**Comment by joekaufmanexpensify** [comment]
[Bumped in Slack](https://expensify.slack.com/archives/C02NK2DQWUX/p1687278577636129)

**Comment by arthurgubaidullin** [comment]
> [Bumped in Slack](https://expensify.slack.com/archives/C02NK2DQWUX/p1687278577636129)

@joekaufmanexpensify What's there? I can't get an invite to slack.

**Comment by arthurgubaidullin** [comment]
@aimane-chnaif Regarding your case. I think this kind of synchronization cannot be solved only on the client. Since device A and device B may lose connection periodically and we are following the offline first strategy, we cannot simply wait for the configurator state changes on the client.

We can use the [lease](https://martinfowler.com/articles/patterns-of-distributed-systems/time-bound-lease.html) pattern:
- Device A requests and receives codes and their validity period, the backend sets lease;
- Device B requests codes and is denied with a timer;
- Device A completes configuration, backend deactivates lease;
- Device A cancels configuration, backend deactivates lease;
- Device B gets an account update.

My solution can be adapted to this pattern.

By redirecting pages. I think what we need is not a redirect, but a redirect with replacement of the current state of the history. Similar to the [replace](https://developer.mozilla.org/en-US/docs/Web/API/Location/replace) method on the web.

At the moment I only have a working demo, I have not integrated it into the project. I wanted to do it in the PR process. It is not profitable for me to do work without result and payment.

Can you clarify if the task is still external or internal?

Thanks.

**Comment by joekaufmanexpensify** [comment]
Still discussing proposals here. Seems like we may be close.

**Comment by joekaufmanexpensify** [comment]
Bumped in Slack to see if we are close to accepting any of these proposals.

**Comment by arthurgubaidullin** [comment]
@aimane-chnaif This is the first time I see such a requirement here. The Contribution Guide says to start coding the solution in the pull request after my solution is selected. Have the rules changed yet? Can you point me to the current version of the rules? I'm a little confused.

**Comment by joekaufmanexpensify** [comment]
Still pending selection of proposal, but we have a few, so should hopefully be close.

**Comment by joekaufmanexpensify** [comment]
[Bumped review in Slack ](https://expensify.slack.com/archives/C02NK2DQWUX/p1687980101894939)

**Comment by joekaufmanexpensify** [comment]
Floated this to callstack as well.

**Comment by koko57** [comment]
Hi I'm Agata from Callstack - expert contributor group - I would like to help to close this issue out.

**Comment by joekaufmanexpensify** [comment]
Great, thanks! @koko57 please let us know once you have an ETA for your proposal.

**Comment by arthurgubaidullin** [comment]
@joekaufmanexpensify Hi. I am not familiar with the company's processes and the guide does not describe the involvement of a person from the agency. Can you explain if a person from the agency has already been selected as a performer or will their proposal be selected on an equal basis with all performers? Thank you very much.

**Comment by koko57** [comment]
@joekaufmanexpensify I've read the whole conversation and have started some investigation here. I've checked @WikusKriek and @hoangzinh solutions but they still have some problems when navigating back. Before I've read all the proposals the first thing that came to my mind was fixing this with a HOC or a hook that will handle the redirect to a proper page (enabled/codes). I also thought about making this one-page just like @bondydaa suggested [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1686755249865569?thread_ts=1686692724.943929&cid=C01GTK53T8Q). What do you guys think about it? 

Just to make sure - we want to redirect user to the settings security page (we want get the user out of enabling 2fa process) when clicking the back button? From any step (s)he's currently on?

About my ETA - I'm not sure if I'll be able to deliver a full proposal today, I'm off on the weekend, but I'll do my best to provide some solution on Monday.

**Comment by koko57** [comment]
> Do your proposals handle the case when user is on any of the steps and does any action on another device logged in same account? i.e.
> 
> * On device A, user is on /verify step and leave that page open. On device B, restart flow and enable 2FA. User tries to enable again on A.
> * On device A, user is on /enabled step and leave that page open. On device B, disable 2FA. User tries to disable again on A
> 

@aimane-chnaif I forgot to ask about the case you mentioned - what is an expected behavior here when dealing with multiple devices?

**Comment by joekaufmanexpensify** [comment]
@koko57 This all sounds great. LMK if there's anything else we can help with as you work on your proposal!

**Comment by deetergp** [comment]
> I also thought about making this one-page just like @bondydaa suggested [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1686755249865569?thread_ts=1686692724.943929&cid=C01GTK53T8Q). What do you guys think about it?

I very much like the idea of it being a one-pager 👍

**Comment by joekaufmanexpensify** [comment]
Not overdue. Pending proposal from @koko57

**Comment by koko57** [comment]
Hi guys! Sorry for the delay. I'm getting closer to finishing the proposal, but I have to check a few more things before I post it. I'll keep you updated.

**Comment by koko57** [comment]
## Proposal @reviewers

### Please re-state the problem that we are trying to solve in this issue.

Secret keys are not displayed when the 2FA page is accessed directly through url. They are also not generated when navigating back from the Verify step of the process.

Additionally, the user should not be able to access all the 2FA steps directly through url.

### What is the root cause of that problem?

The codes are generated when we navigate from the security main page to the step 1 of the 2FA enabling process. They should be generated when the page with codes mounts.

### What changes do you think we should make in order to solve the problem?

Option 1. Current solution - several pages.

Tested @WikusKriek [proposal](https://github.com/Expensify/App/issues/19496#issuecomment-1579064976) once again. It works as expected. Navigating back works fine - I guess I was confused about the navigating back from the Verify step - not sure if we should leave the process and navigate back to the security settings page or back to the codes - whatever we choose can be implemented easily in @WikusKriek solution (for now it's navigating back to the codes page).

Hooks. I think that what arthurgubaidullin meant in his [solution](https://github.com/Expensify/App/issues/19496#issuecomment-1579162504), but to be more specific how do I see this: we can create a hook that will check for the 2FA requirement status and if the process is ongoing (we don’t want to trigger the hook when we are navigating through pages during the process). For the latter we can check if codes are already generated (they are stored in Onyx during the process) or create an additional flag to store in Onyx. The only drawback of this solution is that we have to remember about all the cases when we don’t want to redirect the user. When trying to implement this solution I had some problems with Success and Disable page - I think that it could be solved by adding some additional flag to Onyx or removing these pages and displaying the content conditionally on the Verify (success) and IsEnabled (disabled) pages. This solution is much more complicated that the first one


Option 2. One page

Instead of having each step on a separate route we could display all the steps on the one page and display them accordingly to the current status of the 2FA enabling process.

Instead of a several /two-factor-auth routes we would have a single /two-factor-auth route.

- 2FA enabled -> we display a button to disable 2FA (currently IsEnabledPage)
- 2FA disabled -> we display codes (currently CodesPage) OR a button to start the process. For the second option we could use the IsEnabledPage design by changing only the header, button text and icon, leaving the message as it is.

The wrapper component will hold the state of the process, passing each step will update the state. The current 2FA process state can be also stored in Onyx.

We could navigate between steps by arrows as we do it now by creating a function that changes the current step and passing this to the HeaderWithBackButton.

However this solution would mean a lot of refactor work.

—

The most important change for the both options is moving `Session.toggleTwoFactorAuth(true);` to the CodesPage and removing `TwoFactorAuthActions.clearTwoFactorAuthData();` from the CodesPage useEffect cleanup and calling it when we exit the process (clicking back button or finishing the process successfully). These changes will also fix the https://github.com/Expensify/App/issues/20611 issue.

To sum up: I'd personally go with @WikusKriek solution as it's working great and is the most simple one. It also allow us to leave the several pages logic as we have now.

### What alternative solutions did you explore? (Optional)

N/A

**Comment by joekaufmanexpensify** [comment]
@aimane-chnaif could you please review this when you have a chance?

**Comment by joekaufmanexpensify** [comment]
Bumped in Slack.

**Comment by joekaufmanexpensify** [comment]
@aimane-chnaif confirmed they should complete their review of this proposal today!

**Comment by joekaufmanexpensify** [comment]
[Bumped in Slack ](https://expensify.slack.com/archives/C02NK2DQWUX/p1689105970926719)

**Comment by deetergp** [comment]
After talking it over, we are going to move forward with @koko57 second proposal to make the 2FA process a one-page experience.

**Comment by koko57** [comment]
Thanks! So I'm getting to work! 😃

**Comment by joekaufmanexpensify** [comment]
Great, thanks!

**Comment by koko57** [comment]
Working on my PR 🙂

**Comment by joekaufmanexpensify** [comment]
Sounds great, thanks!

**Comment by joekaufmanexpensify** [comment]
Not overdue, waiting for PR

**Comment by joekaufmanexpensify** [comment]
I'd expect that you'd go back to the codes page. Do you agree @aimane-chnaif, @deetergp ?

**Comment by koko57** [comment]
So I've implemented it that way: 
success page -> back to settings
verify page -> back to codes
but I can change that easily, if necessary

**Comment by koko57** [comment]
@aimane-chnaif bumping for re-review 🙂

**Comment by joekaufmanexpensify** [comment]
PR got merged last week!

**Comment by joekaufmanexpensify** [comment]
Payment due here on 2023-08-30!

**Comment by joekaufmanexpensify** [comment]
Sounds good!

**Comment by joekaufmanexpensify** [comment]
Original upwork job closed, so I created a new one [here](https://www.upwork.com/jobs/~0174aae4bba321c552) to pay @adeel0202 . 

Separately, @sakluger got assigned [a duplicate issue](https://github.com/Expensify/App/issues/25518) to pay @aimane-chnaif for their review in this issue and already hired them there, so we're going to use that job for the purposes of paying @aimane-chnaif.

**Comment by joekaufmanexpensify** [comment]
@adeel0202 offer sent for $250!

**Comment by joekaufmanexpensify** [comment]
@adeel0202 $250 sent and contract ended!

**Comment by joekaufmanexpensify** [comment]
@aimane-chnaif $2,000 sent for the upwork contract you accepted in [this issue](https://github.com/Expensify/App/issues/25518#issuecomment-1701775554), and contract ended.

**Comment by joekaufmanexpensify** [comment]
Closing as this one's all set. Thanks everyone!
